### PR TITLE
Do fault tolerance checks on status object

### PR DESCRIPTION
**Pull request by sbodagala**
# Description

Do fault tolerance checks on FoundationDB status object.

## Type of change

*Please select one of the options below.*

- New feature (non-breaking change which adds functionality)

## Discussion

*Are there any design details that you would like to discuss further?*

No.

## Testing

*Please describe the tests that you ran to verify your changes. Unit tests?
Manual testing?*

Ran operator unit tests locally.

*Do we need to perform additional testing once this is merged, or perform in a larger testing environment?*

Yes, nightly regression tests will need to be run. (But I think nightly tests only run upgrade related tests and we don't test this functionality in those tests, so not sure if running nightly regression tests will help here.)

## Documentation

*Did you update relevant documentation within this repository?*

No (maybe at a later stage).

*If this change is adding new functionality, do we need to describe it in our user manual?*

No (maybe at a later stage).

*If this change is adding or removing subreconcilers, have we updated the core technical design doc to reflect that?*

N/A

*If this change is adding new safety checks or new potential failure modes, have we documented and how to debug potential issues?*

Not sure (maybe we need to document this at a later stage).

## Follow-up

*Are there any follow-up issues that we should pursue in the future?*

No.

*Does this introduce new defaults that we should re-evaluate in the future?*

N/A

### CODE REVIEW

File: internal/foundationdb_status.go [hunk 1]
```diff
@@ -95,3 +95,25 @@ func GetMinimumUptimeAndAddressMap(logger logr.Logger, cluster *fdbv1beta2.Found
...
+func DoFaultDomainChecksOnStatusOutput(status *fdbv1beta2.FoundationDBStatus) bool {
+	// Storage server related check.
+	for _, tracker := range status.Cluster.Data.TeamTrackers {
+		if !tracker.State.Healthy {
+			return false
+		}
+	}
+
+	// Log server related check.
+	for _, log := range status.Cluster.Logs {
```
Comment (johscheuer) [inline internal/foundationdb_status.go:109]: Do you think it make sense to check all logs generations? Or should be the current epoch good enough?

**Review by johscheuer — COMMENTED** [review]
Overall the changes look good, a few suggestion from my side and could we add some basic tests for the new method?

File: internal/foundationdb_status.go [hunk 2]
```diff
@@ -95,3 +95,25 @@ func GetMinimumUptimeAndAddressMap(logger logr.Logger, cluster *fdbv1beta2.Found
 
 	return minimumUptime, addressMap, nil
 }
+
+// DoFaultDomainChecksOnStatusOutput does fault domain checks over the given status object.
+func DoFaultDomainChecksOnStatusOutput(status *fdbv1beta2.FoundationDBStatus) bool {
```
[the remark this answers is not in the document]
Comment (sbodagala) [reply internal/foundationdb_status.go:100]: I assume this function/functionality is needed (and will be used) by the CanI subsystem (as mentioned in the design document).

Regarding writing unit tests: it is a very simple function, not sure if it needs a unit test of its own. But if you insist I can write a unit test (in which case, do we need to manually compose the status input that we will need to give to this function?)

[the remark this answers is not in the document]
Comment (sbodagala) [reply internal/foundationdb_status.go]: Done.

File: internal/foundationdb_status.go (hunk 1 above)
Comment (sbodagala) [reply internal/foundationdb_status.go:114]: I don't have a clear answer right now. I did add "current" to the status structure, and a @todo item about it (so we can revisit this later).

**Pull request closed without merging**